// vite-plugin-framer/index.js
// @ts-check

import { normalizePath, version as viteVersion } from "vite"
import fs from "node:fs/promises"
import path from "node:path"
import picocolors from "picocolors"
import browserslistToEsbuild from "browserslist-to-esbuild"
const { blue } = picocolors

const viteMajor = Number(viteVersion.split(".")[0])
const bundlerOptionsKey = viteMajor >= 8 ? "rolldownOptions" : "rollupOptions"

const configPath = "framer.json"
const docsUrl = "https://framer.com/plugins/open"
const framerIconSvg = `data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIxNiIgaGVpZ2h0PSIxNiIgdmlld2JveD0iMCAwIDE2IDE2Ij48cGF0aCBkPSJNMyAxaDEwdjVIOFptMCA1aDVsNSA1SDNabTUgNXY1bC01LTVaIi8+PC9zdmc+`
const isProductionBuild = process.env.NODE_ENV === "production"

// Keep up-to-date with FramerStudio
const target = browserslistToEsbuild([
    ">0.25% and last 2 years and not dead and not OperaMobile > 1 and not Samsung > 1 and not UCAndroid > 1 and not FirefoxAndroid > 1 and not Android > 1 and not and_qq > 1 and not and_chr > 1 and not ios_saf > 1 and not op_mini all and not kaios > 1",
    "Firefox ESR",
    "unreleased Chrome versions",
    "unreleased Firefox versions",
    "unreleased Safari versions",
])

function replaceCss(html, scriptFilename, scriptCode) {
    const reStyle = new RegExp(`<link([^>]*?) href="[./]*${scriptFilename}"([^>]*?)>`)
    const legacyCharSetDeclaration = /@charset "UTF-8";/
    const inlined = html.replace(
        reStyle,
        (_, beforeSrc, afterSrc) =>
            `<style${beforeSrc}${afterSrc}>${scriptCode.replace(legacyCharSetDeclaration, "")}</style>`,
    )
    return inlined
}

/**
 * Framer Plugin Dev Server Plugin.
 *
 * @returns {import('vite').Plugin}
 */
const framerPlugin = () => {
    let config

    return {
        name: "vite-plugin-framer",
        enforce: "post",
        configureServer() {
            console.log(`\n${blue("➜")} Open your Plugin in Framer: ${blue(docsUrl)}`)
        },
        configResolved(_config) {
            config = _config
        },
        generateBundle: (_, bundle) => {
            if (!isProductionBuild) {
                return
            }

            // Check the output CSS chunks and if they are small we inline them into the HTML instead.
            /**
             * The main HTML has a brotli compressed size of around 1.4 kb, which leaves us around 11 kb to stay below 14 kb: https://web.dev/articles/extract-critical-css#14KB
             * The limit below reflects the non-compressed size, with brotli the real size is around 50% smaller (e.g. for 20 kb, the inlined size becomes <10kb)
             */
            let assetsInlineLimit = 20 * 1024 // 20 kb is enough to inline the default CSS + some authored CSS + some buffer

            const files = { css: [], html: [] }
            const isCSS = /\.css$/
            const isHTML = /\.html$/

            for (const fileName of Object.keys(bundle)) {
                if (isCSS.exec(fileName)) {
                    files.css.push(fileName)
                } else if (isHTML.exec(fileName)) {
                    files.html.push(fileName)
                }
            }

            const bundlesToDelete = new Set()

            for (const name of files.html) {
                const htmlChunk = bundle[name]
                if (htmlChunk.type !== "asset") continue

                let replacedHtml = htmlChunk.source

                for (const filename of files.css) {
                    const cssChunk = bundle[filename]
                    if (cssChunk.type !== "asset") continue

                    const source = cssChunk.source

                    if (source != null && source.length < assetsInlineLimit) {
                        const nextHtml = replaceCss(replacedHtml, cssChunk.fileName, cssChunk.source)
                        if (nextHtml === replacedHtml) continue

                        assetsInlineLimit -= source.length
                        replacedHtml = nextHtml
                        bundlesToDelete.add(filename)
                    }
                }
                htmlChunk.source = replacedHtml
            }

            for (const name of bundlesToDelete) {
                delete bundle[name]
            }
        },
        transformIndexHtml(html) {
            const errorDiv = `
      <div id="framer-environment-error" style="display: none;">
          <svg xmlns="http://www.w3.org/2000/svg" width="26" height="26" viewbox="0 0 20 20"><path d="M5 2.5h10v5h-5Zm0 5h5l5 5H5Zm5 5v5l-5-5Z"/></svg>
          <p>This is a Framer Plugin. Begin your development by opening this plugin within Framer. <a href="${docsUrl}">Learn More</a></p>
      </div>
      `

            const detectionScript = `
      <script>
      document.addEventListener('DOMContentLoaded', () => {
          // Framer Plugin development iframe detection
          if (window.self === window.top) {
              document.getElementById('framer-environment-error').style.display = 'flex';
          }
          // Override Favicon
          var link = document.querySelector("link[rel~='icon']");
          if (!link) {
              link = document.createElement('link');
              link.rel = 'icon';
              document.head.appendChild(link);
          }
          link.href = "${framerIconSvg}";
      });
      </script>
      `

            if (isProductionBuild) {
                // remove `crossorigin` from the script & link tags that are relative
                html = html
                    .replace(/<script type="module" crossorigin src="\//g, '<script type="module" src="/')
                    .replace(/<link rel="stylesheet" crossorigin href="\//g, '<link rel="stylesheet" href="/')
            }

            return html.replace("</body>", `${errorDiv}${detectionScript}</body>`)
        },
        config: () => ({
            assetsInclude: configPath,
            build: {
                target,
                // lightningcss (Vite 8) doesn't support generic ES targets like "es2022".
                cssTarget: target.filter((t) => !/^es\d/.test(t)),
                [bundlerOptionsKey]: {
                    preserveEntrySignatures: false,
                    output: {
                        ...(viteMajor >= 8 && {
                            codeSplitting: {
                                groups: [
                                    {
                                        name: "index",
                                        minShareCount: 2,
                                    },
                                ],
                            },
                        }),
                        // Keep Framer plugin entries as `.mjs`.
                        // https://github.com/framer/company/issues/30016
                        entryFileNames: "[name]-[hash].mjs",
                    },
                },
            },
            // In Vite 8+, Rolldown uses build.target for dep optimization.
            ...(viteMajor < 8 && {
                optimizeDeps: { esbuildOptions: { target } },
            }),
            server: { cors: true },
        }),
        async writeBundle() {
            const src = normalizePath(path.resolve(config.root, configPath))
            const dest = normalizePath(path.resolve(config.build.outDir, configPath))

            await fs.copyFile(src, dest)
        },
    }
}

export default framerPlugin
